### PR TITLE
esp8266: Add per-board configs, following other ports.

### PR DESCRIPTION
**Pull request by mcauser**
Defaults to BOARD=GENERIC.

- [x] Added /boards
- [x] Added /boards/GENERIC
- [x] Added /boards/GENERIC_512K
- [x] Moved ld files to /boards
- [x] Remove 512k from `Makefile`
- [x] Remove `mpconfigport_512k.h` - it's now in /boards/GENERIC_512K
- [x] Update the `README.md` - 512KB FlashROM version
- [x] Add other boards to test in `.travis.yml`

For later:
* Update whatever builds the 512k firmware on http://micropython.org/download
* Add OTA 1M board definition

### CODE REVIEW

**Comment by mcauser** [comment]
Inspired by https://github.com/micropython/micropython/pull/4991

File: ports/esp8266/Makefile [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
+BUILD ?= build-$(BOARD)
+
+BOARD_DIR ?= boards/$(BOARD)
+ifeq ($(wildcard $(BOARD_DIR)/.),)
+$(error Invalid BOARD specified: $(BOARD_DIR))
+endif
+
 include ../../py/mkenv.mk
 
 # qstr definitions (must come before including py.mk)
```
Comment (jimmo) [inline ports/esp8266/Makefile]: To match other ports, can you also add:

`-include $(BOARD_DIR)/mpconfigboard.mk`

Note the `-` on the front to make it optional.

Although it won't be used for the generic board, it's possible another board might want it (e.g. to reference additional code outside the board dir (e.g. extmod) or to change compiler/linker flags).

**Comment by mcauser** [comment]
Failing when trying to load `build/main.o` in `/ports/esp8266/esp8266_common.ld`
Should be `build-GENERIC/main.o`

**Comment by mcauser** [comment]
So, what are the most popular boards?
* Adafruit Feather HUZZAH
* Adafruit HUZZAH
* NodeMCU DevKit
* SparkFun Thing
* Wemos D1 Mini
* Wemos D1 Mini Lite
* Wemos D1 Mini Pro
* WiFi Witty
* Wio Link

**Comment by nevercast** [comment]
> So, what are the most popular boards
I think we can add those as seperate PRs, right?

@jimmo Are you happy with this PR? Can you say it is reviewed?

**Comment by dpgeorge** [comment]
Looks good to me.

I agree other boards can be added later.  First thing to do would be convert the 512k build to a board.

**Comment by mcauser** [comment]
Had a crack at adding the 512K board. A few things I'm not sure about

a) Can board names start with a number?

b) Not sure about these 3x in mpconfigport_512k.h:
```
#undef mp_import_stat
#undef mp_builtin_open
#undef mp_builtin_open_obj
```
It used to run after `#include <mpconfigport.h>` but now `/boards/512K/mpconfigboard.h` runs before. Add a custom `#define` in the `*board.h` and perform the `#undef's` in the `*port.h` if defined?

c) should esp8266_512k.ld be moved into the /ports/esp8266/boards/512K dir?

Remaining:
* remove 512k from Makefile
* remove mpconfigport_512k.h
* Update the readme - 512KB FlashROM version

**Comment by dpgeorge** [comment]
> Can board names start with a number?

Good question!  Probably best not to.  In this case it could be GENERIC_512K

> b) Not sure about these 3x in `mpconfigport_512k.h`...

Are they needed at all?  What happens with these just removed?

> Should `esp8266_512k.ld` be moved

Maybe moved to `ports/esp8266/boards/` dir, so it can be reused by other 512k boards if needed.

> Should every `#define` in `/ports/esp8266/mpconfigport.h` be wrapped in a `#ifndef` so boards can define them (first)? Or only add when first needed?

Only add when first needed.

**Comment by dpgeorge** [comment]
@mcauser is this ready to review, or do you plan to tick all the boxes in the first post?

**Comment by mcauser** [comment]
@dpgeorge I’ll cleanup the 512k bits first then push the other tasks to separate PRs

**Comment by mcauser** [comment]
Hi @dpgeorge , Over to you for review.

**Comment by mcauser** [comment]
```
$ tools/check_code_size.sh
Old size: 68848 new size: 68852
Validation failure: Core code size increased
The command "tools/check_code_size.sh" exited with 1.
```
Not sure what increased the minimal build by 4 (bytes?).

**Comment by nevercast** [comment]
> Not sure what increased the minimal build by 4 (bytes?).

Did the board or MCU name change ?

**Comment by mcauser** [comment]
@nevercast on the 512K board it did.
```
#define MICROPY_HW_BOARD_NAME "ESP module"
#define MICROPY_HW_BOARD_NAME "ESP module (512K)"
```

**Comment by mcauser** [comment]
Rebased off master, CI now passes 👌

File: ports/esp8266/boards/esp8266_common.ld [hunk 2]
```diff
@@ -142,7 +142,7 @@ SECTIONS
         *lib/utils/interrupt_char.o*(.literal.mp_hal_set_interrupt_char, .text.mp_hal_set_interrupt_char)
         *drivers/bus/*.o(.literal* .text*)
 
-        build/main.o(.literal* .text*)
+        build-*/main.o(.literal* .text*)
```
Comment (dpgeorge) [inline ports/esp8266/boards/esp8266_common.ld]: I think it'll work the same if it's just: `main.o(.literal* .text*)`

File: ports/esp8266/boards/esp8266_common.ld [hunk 3]
```diff
@@ -179,7 +179,7 @@ SECTIONS
         */frozen.o(.rodata.mp_frozen_content) /* frozen modules */
 
         /* for -mforce-l32 */
-        build/*.o(.rodata*)
+        build-*/*.o(.rodata*)
```
Comment (dpgeorge) [inline ports/esp8266/boards/esp8266_common.ld]: Probably can be just: `*.o(.rodata*)`

File: ports/esp8266/mpconfigport.h [hunk 4]
```diff
@@ -10,12 +14,26 @@
 #define MICROPY_ALLOC_PARSE_RULE_INC    (8)
 #define MICROPY_ALLOC_PARSE_RESULT_INC  (8)
 #define MICROPY_ALLOC_PARSE_CHUNK_INIT  (64)
+
+#ifndef MICROPY_PERSISTENT_CODE_LOAD
 #define MICROPY_PERSISTENT_CODE_LOAD (1)
+#endif
```
Comment (dpgeorge) [inline ports/esp8266/mpconfigport.h]: I think with all these options that have guards added in this file, it'd be much simple to just not configure them at all here.  Instead, the mpconfigboard.h file should enable the features it wants.  So in GENERIC_512K/mpconfigboard.h it'd be pretty much empty.  And in GENERIC/mpconfigboard.h it'd enable all the extra features.

Then this file here would be almost unchanged.

**Review by dpgeorge — COMMENTED** [review]
Great work!  Just a few comments inline.

File: ports/esp8266/mpconfigport.h (hunk 4 above)
Comment (mcauser) [reply ports/esp8266/mpconfigport.h]: That would bring it closer in line with the STM32 port, where features are off by default and enabled in board specific mpconfigboard.h's

File: ports/esp8266/boards/esp8266_common.ld (hunk 2 above)
Comment (mcauser) [reply ports/esp8266/boards/esp8266_common.ld]: Fixed.

File: ports/esp8266/boards/esp8266_common.ld (hunk 3 above)
Comment (mcauser) [reply ports/esp8266/boards/esp8266_common.ld]: Fixed.

File: ports/esp8266/mpconfigport.h (hunk 4 above)
Comment (mcauser) [reply ports/esp8266/mpconfigport.h]: Updated as per your recommendations.
Defaults in `/py/mpconfig.h`, with common features enabled in `/ports/esp8266/mpconfigboard.h`, and board specific features enabled in `/ports/esp8266/boards/GENERIC/mpconfigboard.h`.

Comment (dpgeorge) [reply ports/esp8266/mpconfigport.h]: Looks good!

Comment (mcauser) [reply ports/esp8266/mpconfigport.h]: In the STM32 port, there's an additional `mpconfigboard_common.h`, which sets defaults for boards which do not provide enough config.

Comment (dpgeorge) [reply ports/esp8266/mpconfigport.h]: That extra complexity isn't needed here, at least not now.

**Comment by dpgeorge** [comment]
Not sure why the 512k build fails...

**Comment by dpgeorge** [comment]
> Not sure why the 512k build fails...

Looks like it has to do with the changes I suggested to esp8266_common.ld, removing build-*.

Can you please revert that, ie use `build-*/main.o ...` etc?

**Comment by mcauser** [comment]
```
xtensa-lx106-elf-ld: build-GENERIC_512K/firmware.elf section `.irom0.text' will not fit in region `irom0_0_seg'
xtensa-lx106-elf-ld: region `irom0_0_seg' overflowed by 1544 bytes
make: *** [build-GENERIC_512K/firmware.elf] Error 1
```
Undo, Abort, Retry

**Comment by mcauser** [comment]
Rolled back changes to `esp8266_common.ld`. Builds locally now.
```
LINK build-GENERIC_512K/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 498124	   1024	  65960	 565108	  89f74	build-GENERIC_512K/firmware.elf
Create build-GENERIC_512K/firmware-combined.bin
esptool.py v1.2
flash     33088
padding   3776
irom0text 466104
total     502968
md5       06617a4ca7a2046cbf55ddb860564949
```

**Comment by mcauser** [comment]
GENERIC also builds successfully on my local.
```
LINK build-GENERIC/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 586988	   1040	  66520	 654548	  9fcd4	build-GENERIC/firmware.elf
Create build-GENERIC/firmware-combined.bin
esptool.py v1.2
flash     33104
padding   3760
irom0text 554968
total     591832
md5       e75afe51f0f8e91323cc8df065baaf33
```

**Comment by mcauser** [comment]
Found an old ESP-01 with 512K flash. Deployed `GENERIC_512K` to it. Works!

```
MicroPython v1.11-422-g2941879e0-dirty on 2019-10-10; ESP module (512K) with ESP8266
Type "help()" for more information.
>>> import micropython
>>> micropython.mem_info()
stack: 2112 out of 8192
GC: total: 37952, used: 688, free: 37264
 No. of 1-blocks: 14, 2-blocks: 7, max blk sz: 5, max free sz: 2325
>>> import gc
>>> gc.mem_free()
37104
>>> gc.mem_alloc()
960
```

**Comment by dpgeorge** [comment]
Really great work, thank you very much!

Rebased, squashed and merged in 53a9b45da1f7f625e273125c259bad7793f17e47 and 3117fde407064cd6e51c4f4bdd35ac9ca3cca20f

**Comment by nevercast** [comment]
Nice work!

**Pull request closed without merging**